fe/fe-core/src/main/java/com/starrocks/qe/Coordinator.java

```diff
@@ -1523,6 +1523,32 @@ public void updateFragmentExecStatus(TReportExecStatusParams params) {
             profileAlreadyReported = true;
         }
 
+        // Update runtime profile when query is still in process.
+        //
+        // We need to export profile to ProfileManager before update this profile, because:
+        // Each fragment instance will report its state based on their on own timer, and basically, these
+        // timers are consistent. So we can assume that all the instances will report profile in a very short
+        // time range, if we choose to export the profile to profile manager after update this instance's profile,
+        // the whole profile may include the information from the previous report, except for the current instance,
+        // which leads to inconsistency.
+        //
+        // So the profile update strategy looks like this: During a short time interval, each instance will report
+        // its execution information. However, when receiving the information reported by the first instance of the
+        // current batch, the previous reported state will be synchronized to the profile manager.
+        if (!execState.done) {
+            long now = System.currentTimeMillis();
+            long lastTime = lastRuntimeProfileUpdateTime.get();
+            if (topProfileSupplier != null &&
+                    connectContext != null &&
+                    connectContext.getSessionVariable().isEnableProfile() &&
+                    now - lastTime > connectContext.getSessionVariable().getRuntimeProfileReportInterval() * 1000L &&
+                    lastRuntimeProfileUpdateTime.compareAndSet(lastTime, now)) {
+                RuntimeProfile profile = topProfileSupplier.get();
+                profile.addChild(buildMergedQueryProfile(null));
```

Comment:
Does `buildMergedQueryProfile` need to be protected by the coordinator lock? Multiple `updateFragmentExecStatus` requests and  the fetching profile request from the client may concurrently access to the query profile.

Reply:
There's a CAS operation, concurrent modifications can be well prevented.